File: Lab5_middleware/SmartHome/src/main/java/server/devices/LightBulb.java
package server.devices;

import SmartHome.*;
import com.zeroc.Ice.Current;

class LightBulb implements SmartHome.LightBulb {
    private final DeviceInfo deviceInfo;
    private LightBulbState lightbulbState;

    LightBulb(DeviceInfo info) {
        this.deviceInfo = info;
    }

    @Override
    public DeviceInfo getDeviceInfo(Current current) {
        return deviceInfo;
    }

    @Override
    public LightBulbState getLightBulbState(Current current) {
        return lightbulbState;
    }

    @Override
    public Color getColor(Current current) throws DeviceTurnedOffError {
        if (!lightbulbState.tunredOn) {
            throw new DeviceTurnedOffError("LightBulb " + deviceInfo.name + " is turned off");
        }
        return lightbulbState.color;
    }

    @Override
    public void setColor(Color newColor, Current current) throws DeviceTurnedOffError, InvalidColorError {
        if (!lightbulbState.tunredOn) {
            throw new DeviceTurnedOffError("LightBulb " + deviceInfo.name + " is turned off");
        }
        lightbulbState.color = newColor;
    }

    @Override
    public void turnOn(Current current) {
        lightbulbState.tunredOn = true;
    }

    @Override
    public void turnOff(Current current) {
        lightbulbState.tunredOn = false;
    }
}

File: Lab5_middleware/SmartHome/client_python/src/client/devices/lightbulb.py
from devices import devices_utils

colors = [
    "white",
    "red",
    "green",
    "blue",
]


def turn_on(lightbulb):
    lightbulb.turnOn()


def turn_off(lightbulb):
    lightbulb.turnOff()


def get_color(lightbulb):
    print("Current color: {}".format(lightbulb.getColor()))


def get_lightbulb_state(lightbulb):
    print("Current state: {}".format(lightbulb.getLightBulbState()))


def get_device_info(lightbulb):
    print("Device info: {}".format(lightbulb.getDeviceInfo()))


def set_color(lightbulb):
    print("Available color")
    for color in colors:
        print(color)
    color = input("new color: ")
    lightbulb.setColor(color)
    get_color(lightbulb)


operations = {
    "info": get_device_info,
    "help": help,
    "turnOn": turn_on,
    "turnOff": turn_off,
    "getColor": get_color,
    "setColor": set_color,
    "getState": get_lightbulb_state,
    "exit": None
}


def lightbulb_handler(lightbulb, device_name):
    devices_utils.run_loop(operations, lightbulb, device_name)
